### PR TITLE
Avoid cloning in ASMsxDmx

### PR DESCRIPTION
**Pull request by akva2**
Unnecessary copies has been used earlier due to lazyness, in particular to avoid bare pointer juggling.
Also unify structure of basis establishment between structured and unstructured ASMs, and fixes handling for ASMmxBase::REDUCED_CONT_RAISE_BASIS2  in the 3D LR ASM.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: src/ASM/ASMs3Dmx.C [hunk 1]
```diff
@@ -51,6 +48,15 @@ ASMs3Dmx::ASMs3Dmx (const ASMs3Dmx& patch, const CharVec& n_f)
 }
 
 
+ASMs3Dmx::~ASMs3Dmx ()
+{
+  // these are managed by shared ptrs, make sure base class do not delete them.
+  if (!this->separateProjectionBasis())
+    projB = nullptr;
```
Comment (kmokstad) [inline src/ASM/ASMs3Dmx.C:55]: projB2?

File: src/ASM/ASMs3Dmx.C [hunk 2]
```diff
@@ -121,6 +127,11 @@ bool ASMs3Dmx::write (std::ostream& os, int basis) const
 
 void ASMs3Dmx::clear (bool retainGeometry)
 {
+  // these are managed by shared ptrs, make sure base class do not delete them.
+  if (!this->separateProjectionBasis())
+    projB = nullptr;
```
Comment (kmokstad) [inline src/ASM/ASMs3Dmx.C:132]: projB2?

File: src/ASM/ASMs3Dmx.C [hunk 3]
```diff
@@ -211,21 +222,22 @@ bool ASMs3Dmx::generateFEMTopology ()
...
-      projB = m_basis.front()->clone();
+    if (!projB)
+      if (ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS1 ||
+          ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2 ||
+          ASMmxBase::Type == ASMmxBase::DIV_COMPATIBLE)
+        projB = ASMmxBase::raiseBasis(svol);
+      else if (ASMmxBase::Type == ASMmxBase::SUBGRID)
+        projB = m_basis.front().get();
+      else // FULL_CONT_RAISE_BASISx
+        projB = m_basis[2-geoBasis].get();
```
Comment (kmokstad) [inline src/ASM/ASMs3Dmx.C:233]: Same question on the added if as in the 2D case, and if yes, add the missing braces.

File: src/ASM/LR/ASMu2Dmx.C [hunk 4]
```diff
@@ -220,33 +220,23 @@ bool ASMu2Dmx::generateFEMTopology ()
...
-      if (!projB)
+    if (!projB)
+      if (ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS1 ||
+          ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2 ||
+          ASMmxBase::Type == ASMmxBase::DIV_COMPATIBLE)
+        projB = createLR(*otherBasis);
+      else if (ASMmxBase::Type == ASMmxBase::SUBGRID)
+        projB = m_basis.front();
+      else // FULL_CONT_RAISE_BASISx
         projB = m_basis[2-ASMmxBase::geoBasis];
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.C:234]: Add braces around L227-234 ?

File: src/ASM/LR/ASMu3Dmx.C [hunk 5]
```diff
@@ -208,28 +208,28 @@ bool ASMu3Dmx::generateFEMTopology ()
...
+      else if (ASMmxBase::Type == ASMmxBase::SUBGRID)
         projB = m_basis.front();
-        refB->generateIDs();
-        projB2 = refB;
-      } else
-        refB = projB = std::make_shared<LR::LRSplineVolume>(otherBasis.get());
-    } else {
-      if (!projB)
+      else // FULL_CONT_RAISE_BASISx
         projB = m_basis[2-ASMmxBase::geoBasis];
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3Dmx.C:227]: braces around L220,227 ?

File: src/ASM/LR/ASMu2Dmx.C [hunk 6]
```diff
@@ -220,33 +220,23 @@ bool ASMu2Dmx::generateFEMTopology ()
...
-        ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2 ||
-        ASMmxBase::Type == ASMmxBase::DIV_COMPATIBLE ||
-        ASMmxBase::Type == ASMmxBase::SUBGRID) {
-      Go::SplineSurface* otherBasis = nullptr;
-      if (!projB)
-        otherBasis = ASMmxBase::raiseBasis(tensorspline);
-
-      if (ASMmxBase::Type == ASMmxBase::SUBGRID) {
-        refB = createLR(*otherBasis);
-        if (!projB)
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.C:234]: OK, so for here we checked for existing projB before also, but not similarly for the other classes. Where is the consistency? And does it make most sense to have the check, or is it unnecessary. That's not clear.

**Review by kmokstad — APPROVED** [review]
Assume this one is fine as well, although the logic is hard to follow.

File: src/ASM/ASMs2Dmx.C [hunk 7]
```diff
@@ -45,6 +45,15 @@ ASMs2Dmx::ASMs2Dmx (const ASMs2Dmx& patch, const CharVec& n_f)
 }
 
 
+ASMs2Dmx::~ASMs2Dmx ()
+{
+  // these are managed by shared ptrs, make sure base class do not delete them.
+  if (!this->separateProjectionBasis())
+    projB = nullptr;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:52]: projB2 is (currently) always a separate object so we want it to be deleted by the base class.

File: src/ASM/ASMs2Dmx.C [hunk 8]
```diff
@@ -120,6 +129,11 @@ bool ASMs2Dmx::write (std::ostream& os, int basis) const
 
 void ASMs2Dmx::clear (bool retainGeometry)
 {
+  // these are managed by shared ptrs, make sure base class do not delete them.
+  if (!this->separateProjectionBasis())
+    projB = nullptr;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:134]: same, it's always a separate object we want to be handled by the base class.

File: src/ASM/ASMs2Dmx.C [hunk 9]
```diff
@@ -210,21 +224,22 @@ bool ASMs2Dmx::generateFEMTopology ()
...
-      projB = m_basis.front()->clone();
+    if (!projB)
+      if (ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS1 ||
+          ASMmxBase::Type == ASMmxBase::REDUCED_CONT_RAISE_BASIS2 ||
+          ASMmxBase::Type == ASMmxBase::DIV_COMPATIBLE)
+        projB = ASMmxBase::raiseBasis(surf);
+      else if (ASMmxBase::Type == ASMmxBase::SUBGRID)
+        projB = m_basis.front().get();
+      else // FULL_CONT_RAISE_BASISx
+        projB = m_basis[2-geoBasis].get();
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:235]: in theory yes since user can have specified a separate projection basis in the input file.

File: src/ASM/ASMs3Dmx.C (hunk 1 above)
Comment (akva2) [reply src/ASM/ASMs3Dmx.C:55]: same, always a separate object.

File: src/ASM/ASMs3Dmx.C (hunk 2 above)
Comment (akva2) [reply src/ASM/ASMs3Dmx.C:132]: same, always a separate object.

File: src/ASM/ASMs3Dmx.C (hunk 3 above)
Comment (akva2) [reply src/ASM/ASMs3Dmx.C:233]: same answer.

File: src/ASM/LR/ASMu2Dmx.C (hunk 4 above)
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C:234]: done.

File: src/ASM/LR/ASMu3Dmx.C (hunk 5 above)
Comment (akva2) [reply src/ASM/LR/ASMu3Dmx.C:227]: done.

File: src/ASM/LR/ASMu2Dmx.C (hunk 6 above)
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C:234]: are you reading the diff reversed? :)

it's consistent now, it used to not be consistent. code had grown organically and sometimes i added a check, sometimes not. now they are all doing exactly the same thing.

**Comment by akva2** [comment]
jenkins build this with downstreams please
